### ISSUE TITLE
Stalemate variations

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#6532]
Opening to continue discussions of variations of stalemate combat rules.

> I strongly believe this should apply at the very least to the first 3 cases I made, namely:
>1- Both sides have 0 power.
2- One or both sides have 1+ power, but no side having positive power is able to target any units of the other side.
3- The battle reaches the maximum number of rounds.

> Both sides having power 0 is conceptually really the same as both sides being unable to target each other (if @panther2 can confirm).

> Max rounds is another kind of stalemate, but it is still a stalemate, and called as such by TripleA, in multiple places. If you are positive not to include it, then I urge you to change the name of this option, as `canRetreatOnStalemate` is definitely saying that I can retreat instead of getting stalemated, in every case.

> However, I think it would be great if this option apply to all the first 3 cases I made. Especially the ability to retreat upon reaching the maximum number of combat rounds would be a great option to have (it would also allow having retreat and 1 combat round, which is currently impossible).

_Originally posted by @Cernelius in https://github.com/triplea-game/triplea/pull/6500#issuecomment-633789979_

**Comment by Cernelius** [comment]
Yeah, in the moment the option is called `canRetreatOnStalemate` I believe it ought to work in every case identified by TripleA as stalemate, that, beside naming consistency, I think it would be very good to increase the value of the option itself.

Meanwhile, pos2 should inform gamemakers under what condition the option is effective.

**Comment by Cernelius** [comment]
I know now that it is merged I could simply test it (and probably will), but I'm also wondering if the 0 power stalemate encompasses any case in which both sides have positive strength but no rolls.

**Comment by DanVanAtta** [comment]
@Cernelius looks like it's just power, not rolls. I suspect it probably should be changed to be the product of rolls and power.

Having `canRetreatOnStalemate` work for max rounds is feasible. Are there any potential maps where we suspect we would want the property to be set one way or another? Is a per-unit option the right way to go? I wouldn't want us to add code for the sake of adding code without it actually having a use.

If we don't think there are any intended use cases, it's not too late to rename the property as `canRetreatOnZeroPowerStalemate`

**Comment by DanVanAtta** [comment]
Mulling it over, I think I'd agree that max round stalemate should behave the same way as a zero power stalemate. https://github.com/triplea-game/triplea/pull/6537 makes that update.

**Comment by Cernelius** [comment]
Ok, so, if we assume
- that having only AA attacks is no cause for the battle to continue (but this would need to be decided and documented, as "if all units left in the battle can hit one another only by AA attacks, whether the battle should immediately end in stalemate or continue") AND
- both sides having only units having the `isInfrastructure` option can never cause a stalemate,

then I think all cases of non-timeout stalemate to cover would be: 

At least one unit for either side has not the `isInfrastructure` option AND every unit in the battle (comprising each having the `isInfrastructure` option) has at least one of these characteristics (after anything is applied (support, marine...)):
- Dice strength equal to 0.
- Dice rolls equal to 0.
- If the unit AND any other units on the same side have not the `isDestroyer` option: every unit type of the opposite side is present in the `canNotTarget` option AND/OR has the `canNotBeTargetedBy` option set for the unit AND/OR has the `isInfrastructure` option.
- If the unit AND/OR at least one other units on the same side have the `isDestroyer` option: every unit type of the opposite side is present in the `canNotTarget` option AND/OR has the `isInfrastructure` option.

The `isDestroyer` case would need some testing, to determine if I got everything right, as I believe pos2 documentation is not very clear, since I'm not sure if what documented is saying that a `isDestroyer` will always cancel all `canNotBeTargetedBy` options of the opposite side or only under a property being true and, in this case, which one. @ron-murhammer maybe can clarify, since I believe he made and documented this matter. This is what it is documented in pos2:
`isDestroyer values: cancels out an isSub unit's first strike and submerge abilities, and allows it to be hit by air units if using ww2v3 rules`
`canNotBeTargetedBy values: colon delimited list of units that can't target this unit, ex. air units not being able to target subs`
What I'm not sure is if the `isDestroyer` case needs a property true to cancel the `canNotBeTargetedBy` option. For now, I'm guessing it doesn't (meaning I'm assuming `isDestroyer` always cancels all `canNotBeTargetedBy` options of every unit on the opposite side), and what I said here is based on this assumption (that may be easily wrong).

**Comment by Cernelius** [comment]
>if all units left in the battle can hit one another only by AA attacks, whether the battle should immediately end in stalemate or continue

In my previous formulation, I've discounted AA attacks to keep it simpler, but I believe they should be considered. Meaning that, in any case of power 0 for both sides, if at least one side has any AA attack able to target any unit of the other side, stalemate should not happen.

**Comment by DanVanAtta** [comment]
@Cernelius , are we still deciding what needs to be done here, or is it pretty clear? If decided, would you mind posting a bulleted list of the things that need to be done.

**Comment by Cernelius** [comment]
We are simply at the same point as per my two previous posts.

There, I'm just detailing the fact that, other than running out of maximum combat rounds (that are infinite by default), you can reach a stalemate when either side has at least 1 unit that is not an infrastructure and either side cannot hit the other side (for having 0 power or being unable to target any units on the other side).

I believe the most typical example of stalemate in which both sides cannot hit each other is given in the World War II v6. Reproduction steps:
- Skip until the start of Germans Combat Move.
- Move 2 submarines to 14 Sea Zone.
- End the Combat Move phase.
- If you kill the carrier while at least 1 submarine remains, you are in a condition in which either side cannot hit the other one (stalemate).
- PROBLEM: TripleA gives you the ability to retreat any remaining submarines (when you have 1 or 2 submarines left against 1 fighter alone), but this is an illegal action, because the battle ended in stalemate and submarines cannot retreat on stalemate (only transports can).

@panther2 Is this problem being currently tracked somewhere else?

---

The point at which I understand we are (though I haven't tested much) is that you have addressed the zero power and finite combat round cases, while the cannot-hit-anything case is still a pending problem (in which stalemate doesn't happen, thus the special ability to retreat on stalemate cannot be applied, since everything can already uncorrectly retreat regardless).

**Comment by Cernelius** [comment]
To be clear, in the basic rules sets you reach stalemate only when every side has at least 1 non-infrastructure unit and:

1- Every unit of every side has 0 power.
2- Every unit of every side cannot target any units on the other side.

However, while this would cover all cases to be encountered in such rules sets, addressing the problem this way would not cover all possibilities you may encounter in other games, as, in non-basic games, you may reach a stalemate situation by a combination of them (for example, you may have one side that has 0 power and the other side that cannot target any remaining unit, even though this situation would never happen in the basic games), thus the matter (that must immediately cause a stalemate, before being able to retreat, unless all offending units have the special ability to retreat on stalemate) would be, rather:

-  Every side has at least 1 non-infrastructure unit and every unit of every side has 0 power and/or cannot target any units on the other side.
